File: day4/AOC.py
import re

def parse_input(input):
    nums = []
    numbers = input.split(": ")
    card_nums = numbers[1].split(" | ")
    nums.append(re.findall(r'\d+', card_nums[0]))
    nums.append(re.findall(r'\d+', card_nums[1]))
    return nums

def part1(input):
    nums = parse_input(input)
    total = 0
    winning_nums = {}
    winning_nums = {num: 1 for num in nums[0]}
    found_nums = [num for num in nums[1] if num in winning_nums]
    if len(found_nums)-1 >= 0:
        total =2**(len(found_nums)-1)
        
    return found_nums, total

def part2(input):
    return True

def main():
    input = open("input.txt", "r").read().splitlines()
    part1_sum = 0
    part2_sum = 0
    for i in range(len(input)):
            _, total = part1(input[i])
            part1_sum += total
    print(f"Part 1: {part1_sum}")
    print(f"Part 2: {part2_sum}")

if __name__ == "__main__":
    main()

File: Go/day3/AOC.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

type PartNumbers struct {
	value int
	start int
	end   int
	row   int
}

type Symbol struct {
	value string
	start int
	row   int
}

type Schematic struct {
	PartNumbers []PartNumbers
	Symbols     []Symbol
}

func (s *Schematic) getAdjacentPartNumbers(sym Symbol) []PartNumbers {
	adjacent := []PartNumbers{}
	// seen := map[int]bool{}
	// for _, part := range s.PartNumbers {

	// }
	return adjacent
}

func parseInput(input []byte) Schematic {
	schematic := Schematic{}
	lines := strings.Split(string(input), "\n")
	for row, line := range lines {
		nums := regexp.MustCompile("[0-9]+").FindAllString(line, -1)
		for _, num := range nums {
			index := strings.Index(line, num)
			n, _ := strconv.Atoi(num)
			schematic.PartNumbers = append(schematic.PartNumbers, PartNumbers{n, index, index + len(num), row})
		}

		symbols := regexp.MustCompile("[^.0-9]").FindAllString(line, -1)
		for _, symbol := range symbols {
			// println(symbol)
			index := strings.Index(line, symbol)
			schematic.Symbols = append(schematic.Symbols, Symbol{symbol, index, row})
		}
	}
	return schematic
}

func part1(s *Schematic) int {

	return 0
}

func part2(input []byte) int {
	fmt.Println("Hello World")
	return 0
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	// fmt.Println(input)
	s := parseInput(input)
	part1_sum := part1(&s)
	part2_sum := part2(input)
	fmt.Printf("Part 1: %d\n", part1_sum)
	fmt.Printf("Part 2: %d\n", part2_sum)
}
